### ISSUE TITLE
Conditional visibility not working on advanced Box Shadow Fields + missing toggle

### ISSUE THREAD

**Issue by xxtesaxx** [elegantthemes/create-divi-extension#97]
### Problem Description
It's easy enough to add new Box Shadow fields using the advanced fields but there a re a few issues I'd love to see get improved/fixed.

1.  When adding a custom box shadow field you either have to include "default => []" or add the toggle yourself. Maybe it would be better to always add the toggle unless there is no Box Shadow config using this toggle (e. g. one could want to remove the overall box shadow for the module itself but add 2-3 new box shadow fields for certain elements. Similar to the button module, which has no box shadow and background but a box shadow only for the button itself)

2. Currently conditional display logic does not work. There are still field in place for the old way (depends_show_if) though so my wish would be to let us add the show_if and show_if_not options and show/hide those fields.

3. Document what the "custom_style" is about. The docs say "Use custom styles generation instead of the default" but what does that mean?

**Comment by etstaging** [comment]
### QA Checklist
<strong>Reporter</strong>
- [ ] Helpful title used &nbsp;<img src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" width="16" height="16"/>
  * ***Helpful title format:*** `Scope(s) :: One Sentence To Describe The Issue`
  * ***Example:*** `VB :: Floating Toolbar :: Jumps To Top Of Page`
- [ ] Detailed description <img src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" width="16" height="16"/>
  * ***Description should usually be 150+ characters.***
- [ ] Steps to reproduce included <img src="https://assets-cdn.github.com/images/icons/emoji/unicode/274c.png" width="16" height="16"/>
- [ ] Support Ticket URL included
- [ ] Image and/or gif

<strong>Developer</strong>
- [ ] Steps to reproduce proven solved
- [ ] Test included and coverage acceptable and all Tests passing?
- [ ] PR description has layman's description of problem and solution and "Areas Affected" listed/explained?
- [ ] PR branch includes latest commits from master branch (at the time of checking)?
- [ ] Estimated points on issue
- [ ] Proper code documentation and conventions?
- [ ] Proper RTL appearance
- [ ] i18n used if applicable
- [ ] Ajax/Form used? Nonce used? Perms checked? All user input sanitized and validated? Intentional unsanitized/unescaped functions used?
- [ ] Database saved/updated? Data sanitized and validated?
- [ ] Variable output properly escaped and validated?

<strong>Peer Review</strong>
- [ ] PR branch includes latest commits from master branch (at the time of checking)?
- [ ] Steps to reproduce proven solved
- [ ] Affected areas are tested and ok
- [ ] Proper code documentation and conventions
- [ ] Proper RTL appearance
- [ ] i18n used if applicable
- [ ] Test(s) included and coverage acceptable
- [ ] Ajax/Form used? Nonce used? Perms checked? All user input sanitized and validated? Intentional unsanitized/unescaped functions used?
- [ ] Database saved/updated? Data sanitized and validated?
- [ ] Variable output properly escaped and validated?

<strong>Quality Assurance</strong>
- [ ] All tests passed on CircleCI.
- [ ] Steps to reproduce passed, the fix works.
- [ ] No errors in console or output 
- [ ] Affected areas are tested and ok
- [ ] The fix does not change a customer's website in any surprising or unintended way
  - [ ] If Divi's design <em>is</em> adjusted, settings are migrated gracefully and invisibly

<strong>Design Review</strong>
- [ ] UI is to spec
- [ ] UX is to spec
- [ ] Design is to spec
- [ ] No surprising visual changes to a customer's website will be experienced after updating
  - [ ] Any/all visual changes are migrated gracefully

<strong>Dev Review</strong>
- [ ] Test(s) included and coverage acceptable
- [ ] Ajax/Form used? Nonce used? Perms checked? All user input sanitized and validated? Intentional unsanitized/unescaped functions used?
- [ ] Database saved/updated? Data sanitized and validated?
- [ ] Variable output properly escaped and validated? 
- [ ] i18n used if applicable
- [ ] Implementation acceptable, DRY, Scaleable, future safe
- [ ] Best practices used, cleverness and grokage, ES6 use, etc.

**Comment by jcdotnet** [comment]
Hi @xxtesaxx, I'm having trouble with the conditional logic too. I've added both an overall box shadow field and a custom box shadow field and I want to hide the first one if not background has been selected. 

I mean, the advanced box shadow field depends on another advanced field (background). 

```
'background' 	=> array(
	'css'      	=> array(
	    'main' 	    => "{$this->main_css_element} .main"
	), 
	'settings' 	=> array(
		'color'	=> 'alpha'
	)			
),
'box_shadow' => array(				
	'default' => array(
		'label'    	=> esc_html__( 'Main', 'plugin_text_domain' ),
		'css' => array(
			'main' => "{$this->main_css_element} .main",						
		),
	),
	'test' => array (
		'label'    	=> esc_html__( 'Test', 'plugin_text_domain' ),
	        'css' => array(
			'main' => "{$this->main_css_element} .test",						
		),					
	),
),
```

I've tried the following with no luck:
```
'default' => array (
    	'show_if_not' => array( 'background_color' => ''),				
),
```
I've checked the background_color field and it's empty when not background has been selected, so the default box shadow should be hidden in that case.

Can advanced fields depend on other advanced fields? If so, how can it be achieved?  Could I use depends_show_if in this scenario?

Hope @lots0logs or someone else can support here, thank you.

**Comment by VladET** [comment]
@jcdotnet Yes, they can. You can use `'depends_on'` and `'depends_show_if_not'`/`'depends_show_if'` to show/hide an option based on the value of another option. In your case:

```php
'depends_on'          => array( 'background_color' ),
'depends_show_if_not' => '',
```

`'show_if'`/`'show_if_not'` determines if the CSS will or will not output based on the values of certain settings. Basically, the option will still be visible but the style will only output if the given condition is true.

https://www.elegantthemes.com/documentation/developers/divi-module/advanced-field-types-for-module-settings/#box-shadow-box_shadow

`show_if (array)` — **Output styles** if certain settings have certain values
`show_if_not (array)` — **Output styles** if certain settings do not have certain values

**Comment by jcdotnet** [comment]
Thanks @VladET. After adding those parameters the default shadow field option is always hidden, regardless of the background_color value:

```
'background' 	=> array(
	'css'      	=> array(
	    'main' 	    => "{$this->main_css_element} .main"
	), 
	'settings' 	=> array(
		'color'	=> 'alpha'
	)			
),
'box_shadow' => array(				
	'default' => array(
		'label'    	=> esc_html__( 'Main', 'plugin_text_domain' ),
		'css' => array(
			'main' => "{$this->main_css_element} .main",						
		),
                'depends_on'          => array( 'background_color' ),
                'depends_show_if_not' => '',
	),
	'test' => array (
		'label'    	=> esc_html__( 'Test', 'plugin_text_domain' ),
	        'css' => array(
			'main' => "{$this->main_css_element} .test",						
		),					
	),
),
```
Am I missing something?

**Comment by VladET** [comment]
@jcdotnet No, that looks fine. I added these fields to a module and it's working. Here is a screenshot:

https://screencast-o-matic.com/watch/cF1jX2Fc8z

If you can share a .zip with your module I can take a look.

**Comment by jcdotnet** [comment]
Working for me too, the code above is fine. If someone wants to make an advanced field dependent on another one can follow this example. Thank you :)

**Comment by xxtesaxx** [comment]
Hmm odd, I tried the depends_on and depents_show_if but it never showed up. But maybe I made a mistake e. g. didn't use array for depends_on or so. Anyways, I think since show_if is the new way to go, maybe you can migrate the setting at some point to get rid of the old depends_on :)

**Comment by jcdotnet** [comment]
@xxtesaxx it seems to work only in the visual builder. I've just changed to the PB and the option never shows up either. I'm going to test it up later and I will let you know if there is something wrong.

**Comment by lots0logs** [comment]
RE: 1, In official modules you'll notice that settings for specific elements inside of a module appear together in their own toggle group. So instead of having all box shadow settings under the same group you would have the main setting under Box Shadow and then the others under the group for the element they apply to. Of course that's simply a design decision on our part and so I don't see any reason not to improve it so that your use case works as expected :smiley: 

RE: 2, This was an oversight on our part. The `show_if/not` params for the box shadow field type should have been renamed due to the obvious conflict. We'll address it for sure.

RE 3, I'll have to look into it and get back to you because it looks like it doesn't do what I've implied in the documentation :sweat_smile:

**Comment by IvanChiurcci** [comment]
@lots0logs any progress regarding `show_if ` issue? 

I have multiple box-shadow fields which make the Design tab look cluttered if I use multiple Box Shadow toggles(separate for each element). So, I added them all into a single Box Shadow toggle and want to show conditionally using a `select` option.

I tried `depends_on/depends_show_if` and it works fine in VB, I am able to select each element one by one and apply the box shadow and all elements preserve their box shadows if I select another one(CSS is applied to all elements). But on front end the box shadow is being applied to the last selected and saved element only.

Any possibility to only show/hide fields without affecting the CSS?

**Comment by lots0logs** [comment]
@IvanChiurcci I'm not sure that your issue is the same as the one's reported here. We will have to look into it. We'll be taking care of many of the high priority issues for 3rd-party api in the upcoming backend builder update. This issue is on that list :)

**Comment by lots0logs** [comment]
~3 weeks

**Comment by fikrirasyid** [comment]
@xxtesaxx @jcdotnet @IvanChiurcci @lots0logs @VladET thank you for your contribution. We've able to locate the cause of the issue and submit a patch for it. The patch is now under review process.

**Comment by IvanChiurcci** [comment]
> @lots0logs any progress regarding `show_if ` issue?
> 
> I have multiple box-shadow fields which make the Design tab look cluttered if I use multiple Box Shadow toggles(separate for each element). So, I added them all into a single Box Shadow toggle and want to show conditionally using a `select` option.
> 
> I tried `depends_on/depends_show_if` and it works fine in VB, I am able to select each element one by one and apply the box shadow and all elements preserve their box shadows if I select another one(CSS is applied to all elements). But on front end the box shadow is being applied to the last selected and saved element only.
> 
> Any possibility to only show/hide fields without affecting the CSS?

@fikrirasyid @lots0logs Unfortunately, looks like the issue I was referring to still hasn't been resolved. I gave up on my idea back then(16 months ago) because it didn't work and now I needed to implement the same functionality for my new project(using "show_if" and color fields) but the issue is the same, the props retain their values in VB and I am able to apply colors to all elements one by one but only as long as I don't save&refresh the builder. If I save only the last set prop value is saved.

**Comment by madjennsy** [comment]
Hi everyone, I would like to know when will this issue be fixed? values are not retained if I add "depends_on" and "depends_show_if" to my module fields after saving the builder. Is there a temporal hack until this issue is fixed?

**Comment by madjennsy** [comment]
Update on the bug:

If I use both 'depends_show_if' and 'depends_on' values are not retained on save. Although if I use only 'depends_show_if', values are saved.
